src/altinn-app-frontend/src/features/form/containers/Form.tsx

```diff
@@ -143,8 +140,34 @@ export function Form() {
     }
   }, [layout]);
 
+  /** history in form **/
+  const dispatch = useAppDispatch();
+  const match = useRouteMatch<string>('/instance/:partyId/:instanceGuid');
+  const matchUrl = match?.url || '';
+  const history = useHistory();
+  React.useEffect(() => {
+    if (currentLayout) {
+      const getPageId = () =>
+        history.location?.pathname.replace(`${matchUrl}/`, '');
+      const urlPageId = getPageId();
+      if (!urlPageId || urlPageId === matchUrl) {
+        history.replace(`${matchUrl}/${currentLayout}`);
+      } else if (`${currentLayout}` !== urlPageId) {
+        history.push(`${matchUrl}/${currentLayout}`);
+      }
+      return history.listen((_location, action) => {
+        if (action === 'POP') {
+          const pageId = getPageId();
+          if (pageId && pageId !== currentLayout) {
+            dispatch(FormLayoutActions.updateCurrentView({ newView: pageId }));
+          }
+        }
+      });
+    }
+  }, [dispatch, currentLayout, matchUrl, history]);
```

Comment:
I would suggest to extract this to a custom hook. That should make it easier to test this in isolation, reduce the cognitive load when reading the Form component, and it might help with understanding the meaning behind this with a descriptive name.

```suggestion
  const {matchUrl} = useHistoryReplaceWhenUrlMatchInstance({currentLayout}); // This name could probably be even better, it was just what I could think of at the top of my head 😅
```


Then extract the content to a separate hook:
```typescript
	
export const useHistoryReplaceWhenUrlMatchInstance = ({currentLayout}) => {
  const dispatch = useAppDispatch();
  const match = useRouteMatch<string>('/instance/:partyId/:instanceGuid');
  const matchUrl = match?.url || '';
  const history = useHistory();

  useEffect(() => {
    if (currentLayout) {
      const pageId = history.location?.pathname.replace(`${matchUrl}/`, '');
      const isOnRootPage = !pageId || pageId === matchUrl;

      if (isOnRootPage) {
        history.replace(`${matchUrl}/${currentLayout}`);
      } else if (`${currentLayout}` !== pageId) {
        history.push(`${matchUrl}/${currentLayout}`);
      }

      return history.listen((_location, action) => {
        const isBrowserBackOrForwardPressed = action === 'POP';

        if (isBrowserBackOrForwardPressed) {
          if (pageId && pageId !== currentLayout) {
            dispatch(FormLayoutActions.updateCurrentView({ newView: pageId }));
          }
        }
      });
    }
  }, [dispatch, currentLayout, matchUrl, history]);

  return {matchUrl};
}


  

```